task/aws/client/client.go

```diff
@@ -0,0 +1,96 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+
+	"terraform-provider-iterative/task/universal"
+	"terraform-provider-iterative/task/universal/ssh"
+)
+
+func New(ctx context.Context, cloud universal.Cloud, tags map[string]string) (*Client, error) {
+	region := string(cloud.Region)
+	regions := map[string]string{
+		"us-east":  "us-east-1",
+		"us-west":  "us-west-1",
+		"eu-north": "eu-north-1",
+		"eu-west":  "eu-west-1",
+	}
+
+	if val, ok := regions[region]; ok {
+		region = val
+	}
+
+	config, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
+	if err != nil {
+		return nil, err
+	}
+
+	c := new(Client)
+	c.Cloud = cloud
+	c.Region = region
+	c.Tags = cloud.Tags
+
+	c.Config = config
+
+	c.Services.EC2 = ec2.NewFromConfig(config)
+	c.Services.S3 = s3.NewFromConfig(config)
+	c.Services.STS = sts.NewFromConfig(config)
+	c.Services.AutoScaling = autoscaling.NewFromConfig(config)
+	return c, nil
+}
+
+type Client struct {
+	Cloud  universal.Cloud
+	Region string
+	Tags   map[string]string
+
+	Config   aws.Config
+	Services struct {
+		EC2         *ec2.Client
+		S3          *s3.Client
+		STS         *sts.Client
+		AutoScaling *autoscaling.Client
+	}
+}
+
+func (c *Client) GetKeyPair(ctx context.Context) (*ssh.DeterministicSSHKeyPair, error) {
+	credentials, err := c.Config.Credentials.Retrieve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return ssh.NewDeterministicSSHKeyPair(credentials.SecretAccessKey, credentials.AccessKeyID)
+}
+
+func (c *Client) DecodeError(ctx context.Context, encoded error) error {
```

Comment:
Not in `./task`, which is self–contained and just part of this repository for practical reasons.

Reply:
>  Not in ./task, which is self–contained and just part of this repository for practical reasons.

Why? what do you mean self- contained? why should not use same aws functions like machine?

Reply:
By self–contained, I mean that the `task` directory is not tied to the legacy implementation in any way.

Reply:
If you do not consider it to do it its fine, thats why I marked it as technical debt, for you to open an issue with the PR

Reply:
> _What is the difficulty of extracting one single function in a module and use it?_

Because of API responsibility. I would not like to create any relationship between the legacy `machine` resource and the new `task` resource until it's more mature, lest slowing down the development process because of interoperatibility considerations.

> _I do not fully understand why we are treating Task as a separate thing_

The experimental `task` library has been built as a standalone, reusable component.

Reply:
> Because of API responsibility. I would not like to create any relationship between the legacy machine resource and the new task resource until it's more mature, lest slowing down the development process because of interoperatibility considerations.

Its just a decode function but whatever

Reply:
https://github.com/iterative/terraform-provider-iterative/pull/237#issue-1048177975